gen3workflow/routes/s3.py

```
@@ -133,6 +132,18 @@ async def s3_endpoint(path: str, request: Request):
         assert credentials, "No AWS credentials found"
         headers["x-amz-security-token"] = credentials.token
 
+    # if this is a PUT request, we need the KMS key ID to use for encryption
+    if config["KMS_ENCRYPTION_ENABLED"] and request.method == "PUT":
+        _, kms_key_arn = aws_utils.get_existing_kms_key_for_bucket(user_bucket)
+        if not kms_key_arn:
+            err_msg = "Bucket misconfigured. Hit the `GET /storage/info` endpoint and try again."
+            logger.error(
+                f"No existing KMS key found for bucket '{user_bucket}'. {err_msg}"
+            )
+            raise HTTPException(HTTP_400_BAD_REQUEST, err_msg)
```

In this scenario, I don't think the user's request itself is not malformed, so a 400 Bad Request response might be misleading. Since the issue comes from a server-side misconfiguration, a `5XX error` would be more appropriate to indicate that the server is temporarily unable to process this `PUT request`. A 503 Service Unavailable, for instance, may convey that the request might succeed after calling `GET /storage/info` to update the configuration.

---

The details of 5xx errors are masked from the end user by revproxy. I chose to return a 400 error so that the user can get the error details that describe how to fix the issue, and because the misconfiguration is caused by the user skipping a step - it's not an unexpected server error